yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,683 @@
+
+```
+kernel_public_inputs.gas_limits := private_kernel_public_inputs.gas_limits
+
+kernel_public_inputs.end.gas_used.da_gas_used :=
+    FIXED_DA_GAS
+    + private_kernel_public_inputs.end.gas_used.da_gas_used
+
+# should be zero
+kernel_public_inputs.end.l2_gas_used :=
+    private_kernel_public_inputs.end.l2_gas_used
+
+# should be zero?
+kernel_public_inputs.end.l1_gas_used :=
+    private_kernel_public_inputs.end.l1_gas_used
+
+# make assertions that gas limits are not exceeded
+
+# purely private transactions cannot pay any fee
+kernel_public_inputs.transaction_fee := 0
+```
+
+# Transaction with public component
+
+### Initialization
+
+If a transaction has a public component, in the `PrivateKernelTailToPublic`, we set:
+```
+public_kernel_public_inputs.gas_limits := private_kernel_public_inputs.gas_limits
+
+public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used :=
+    public_kernel_public_input.gas_limits.teardown_da_gas
+    + FIXED_DA_GAS
+
+public_kernel_public_inputs.end_non_revertible.gas_used.l2_gas_used :=
+    public_kernel_public_input.gas_limits.teardown_l2_gas
+
+public_kernel_public_inputs.end_non_revertible.gas_used.l1_gas_used :=
+    public_kernel_public_input.gas_limits.teardown_l1_gas
+```
+
+Note that **we effectively pre-pay for the teardown function.**
+
+### Happy path interaction with the public VM
+
+The public VM retains the current amount of `daGasLeft`, `l2GasLeft`, and `l1GasLeft` in its [MachineState](../public-vm/state.md#machine-state).
+
+This amount must be initialized for each public kernel iteration as:
+
+```
+daGasLeft := public_kernel_public_inputs.da_gas_limit 
+             - public_kernel_public_inputs.end_non_revertible.da_gas_used
+             - public_kernel_public_inputs.end.da_gas_used
+
+# analogous for l2GasLeft and l1GasLeft
+```
+
+At the end of execution, the public VM will have updated `daGasLeft`, `l2GasLeft`, and `l1GasLeft` values.
+
+This value will be used by the public kernel to update the `da_gas_used`, `l2_gas_used`, and `l1_gas_used` in the `PublicAccumulatedData` struct corresponding to the current phase:
+
+```
+# if in setup 
+public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used := 
+        public_kernel_public_inputs.gas_limits.da_gas_limit
+          - daGasLeft
+          - public_kernel_public_inputs.end.gas_used.da_gas_used
+
+# if in app logic
+public_kernel_public_inputs.end.gas_used.da_gas_used := 
+        public_kernel_public_inputs.gas_limits.da_gas_limit
+          - daGasLeft
+          - public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used
+
+# analogous for l2_gas_used and l1_gas_used
+```
+
+### Global gas limits
+
+The mechanics above imply that there will be a single global gas limit (for each gas dimension) for all enqueued public calls. However, users will still be able to specify individual gas limits for each *nested* call within an enqueued public call.
```

Comment:
> users will still be able to specify individual gas limits for each *nested* call within an enqueued public call

How do they do this? Are there parameters that are baked into the "call" opcodes of the avm? By "users" do you mean the smart contract developer who designs how the nested call is called (and therefore how much gas to send to it)?

Reply:
Wait, so if I have a tx with two enqueued public function calls A1 and B1:

- I specify a global gas limit of 100 for all enqueued public function calls.

But then we're saying I can specify individual gas limits for each nested function call? Is it the caller of the function who can specify these, or the smart contract developer?

So if A1 calls A1 and B1 calls B2, and I (or the functions A1 & B1?) can somehow specify gas limits of:
- A1: 50 gas, but it uses 20 gas.
- A2: 25 gas, but it uses 20 gas.
- B1: 50 gas, but it uses 20 gas.
- B2: 30 gas, but it uses 20 gas.

Would my tx succeed or fail?

Reply:
I understand your example working as:

- A1 starts with 100 gas, nobody can give this a gas limit different from the global tx limit. 
- A1 calls A2 and the developer specifies a limit of 25 gas. If it tries to use more it will revert and can be caught by A1, consuming only 25. 
- A1 itself uses 20 gas. So, in your example, both A1 and A2 use 20 each, 40 total.
- B1 now has an effective limit of 60, again neither the developer or the user can set this themselves
- B1 calls B2 with a limit of 30, They both use 20 gas which does not breach any limit.

80 gas used in total.